### ISSUE TITLE
JAX-RS validation issue with interface methods with default implementation

### ISSUE THREAD

**Issue by Meemaw** [quarkusio/quarkus#13029]
**Describe the bug**
After trying to upgrade to Quarkus 1.9.x (1.9.1) I've seen there were some changes (regressions?) to the validation behaviour. It seems that the validation does not work for interface methods with default implementation)

**Expected behavior**
JAX-RS validation should work for interface methods with default implementation.

**To Reproduce**

BookResource.java

```java
package org.acme.validation;

import javax.validation.constraints.NotBlank;
import javax.ws.rs.CookieParam;
import javax.ws.rs.GET;
import javax.ws.rs.Path;
import javax.ws.rs.PathParam;
import javax.ws.rs.Produces;
import javax.ws.rs.core.MediaType;

@Path("/books")
public interface BookResource {

  @GET
  @Produces(MediaType.TEXT_PLAIN)
  @Path("{user}")
  String helloUser(@PathParam("user") String user);

  @GET
  @Produces(MediaType.TEXT_PLAIN)
  default String hello(@NotBlank(message = "Required") @CookieParam("user") String user) {
    System.out.println("HELLO WORLD: " + user);
    return helloUser(user);
  }

  @GET
  @Produces(MediaType.TEXT_PLAIN)
  @Path("works")
  String thisOneWorks(@NotBlank(message = "Required") @CookieParam("user") String user);
}
```

BookResourceImpl.java

```java
package org.acme.validation;

public class BookResourceImpl implements BookResource {

  @Override
  public String helloUser(String user) {
    return "hello " + user;
  }

  @Override
  public String thisOneWorks(String user) {
    return "hello " + user;
  }
}
```

Visit http://localhost:8080/books and you should see:
```sh
hello null
```

Visit http://localhost:8080/books/works and you should see:
```sh
[PARAMETER]
[thisOneWorks.user]
[Required]
[]
```

**Environment (please complete the following information):**
 - Quarkus version or git rev: 19.1

**Comment by Meemaw** [comment]
Any idea if this can be fixed on Quarkus side? @gsmet

**Comment by Sanne** [comment]
thanks for bringing it to my attention @Meemaw ! Sorry, I think we hadn't noticed it - the label will help prioritizing, but I'll leave that to @gsmet to evaluate and schedule.

**Comment by Meemaw** [comment]
Any plan to fix this @gsmet ?

**Comment by gsmet** [comment]
Sorry I missed this one.

@mkouba do we support interceptors on interface default methods with ArC?

AFAICS, the annotation transformer I have in HV is not called for the default method of the interface so I wonder if it's supported at all. If it's not I should probably raise an error.

**Comment by Meemaw** [comment]
Let me just mention that this works in Quarkus v1.6.1

**Comment by gsmet** [comment]
Yeah, that's a bit weird. I don't see the annotation transformer called for the default method. I'll try to compare with the behavior in 1.6.1 later today.

**Comment by gsmet** [comment]
Yeah, so, I think the problem is in ArC and more precisely at this line: https://github.com/quarkusio/quarkus/blob/master/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Methods.java#L205 .

The last parameter being `true` makes all the methods level annotations ignored.

I have  no idea why this parameter is true in this case but it has been introduced precisely here: https://github.com/quarkusio/quarkus/commit/fa49c9d26308707c46c109a3e00ee9075dcdc762#diff-f05a30ce973915337c900f71476c739f710d3fcc1ab6d148bbb418755a2bd5a9 .

@mkouba I think we will need you to shed some light on this change, it looks suspicious.

**Comment by mkouba** [comment]
> do we support interceptors on interface default methods with ArC?

@gsmet Yes, we do support interceptors on default methods.

WRT the change - it's described in the [issue description](https://github.com/quarkusio/quarkus/issues/11704#issue-687947582). CDI beans cannot inherit interceptor bindings from the implemented interfaces.

**Comment by gsmet** [comment]
@mkouba even for a default method?

**Comment by mkouba** [comment]
AFAIK the relevant specs do not mention default methods at all.

CC @manovotn @Ladicek

**Comment by gsmet** [comment]
The use case presented here has a case IMHO. It looks like something we should support somehow.

**Comment by manovotn** [comment]
> AFAIK the relevant specs do not mention default methods at all.

Yeah, interceptors spec doesn't mention default method which means standard rules apply and that's precisely what Martin said. Basically, if you want to intercept default method, you have to use a class-level interceptor binding on the implementing class.

I agree it is somewhat weird but it's per spec and what's worse - changing the inheritance rule to apply to interfaces would break existing apps.

Maybe we should also file an issue against interceptors specification to get a spec level opinion on what it should behave like?

**Comment by mkouba** [comment]
> Maybe we should also file an issue against interceptors specification to get a spec level opinion on what it should behave like?

+1

**Comment by manovotn** [comment]
I have created https://github.com/eclipse-ee4j/interceptor-api/issues/84 to track this on a spec-level.

**Comment by gsmet** [comment]
So basically we can't support constrained default methods for JAX-RS? Or is there a clever workaround we could apply?

**Comment by mkouba** [comment]
> So basically we can't support constrained default methods for JAX-RS? Or is there a clever workaround we could apply?

I'm not aware of _any_ workaround.

**Comment by gsmet** [comment]
OK. That's what I thought. From a user perspective, it's really not good.

I think I will have to throw an error in this case but I need to figure out how I will be able to handle this case and only this one.

**Comment by mkouba** [comment]
> OK. That's what I thought. From a user perspective, it's really not good.

I agree :-(. Unfortunately, there seems to be a spec mismatch...

**Comment by manovotn** [comment]
> Or is there a clever workaround we could apply?

To sum up, the current way it works is also against what CDI spec says (normally, you wouldn't be able to inherit member-level metadata). Quarkus uses `io.quarkus.hibernate.validator.deployment.MethodValidatedAnnotationsTransformer` to work around this by inspecting hierarchy and then adding bindings programmatically.

So initially I thought you could twist this annotation transformer and make it work for default methods but that's not the case. By virtue of being default, you don't have them on the implementing class and so you cannot add the binding to them even if you identified them. So you would need to implement a bytecode transformation and override the method in impl class (-10 to that, I know) at which point you could add bindings. Other than that I don't see an option because we basically lack means of adding a member-level binding to a default method.

**Comment by mkouba** [comment]
> So you would need to implement a bytecode transformation and override the method in impl class...

Yes, this could work but seems to be a "fragile" solution, also it would be a mess if a more complex class hierarchy is used...

**Comment by gsmet** [comment]
Well, if you put the spec aside, I think the only issue is where I pointed it here: https://github.com/quarkusio/quarkus/issues/13029#issuecomment-735232180 .

If we consider default methods for bindings, then I think it would all work.

**Comment by Meemaw** [comment]
In case you decide not to support this, I think error should be thrown so users don't expect this to work, while in reality in doesn't.

**Comment by yrodiere** [comment]
This got reported again today as #50660.

@manovotn @mkouba @Ladicek what do you think of Guillaume's suggestion to implement a solution?

> Well, if you put the spec aside, I think the only issue is where I pointed it here: [#13029 (comment)](https://github.com/quarkusio/quarkus/issues/13029#issuecomment-735232180) .
> 
> If we consider default methods for bindings, then I think it would all work.

Note if the spec is a problem, we could have only some interceptors follow this behavior (e.g. Validator's)... ?

**Comment by Ladicek** [comment]
Note that even the JAX-RS specification, which allows annotation inheritance from interfaces, only does that for JAX-RS annotations. No other annotations are supposed to be inherited. See https://jakarta.ee/specifications/restful-ws/4.0/jakarta-restful-ws-spec-4.0#annotationinheritance 

My favorite sentence from that specification is this:

> For consistency with other Jakarta EE specifications, it is recommended to always repeat annotations instead of relying on annotation inheritance.

This is what I would recommend people to do.
